newlib/libc/sys/hermit/fcntl.c

```diff
@@ -49,3 +50,29 @@ int open(const char *path, int oflag, ...) {
 
     return ret;
 }
+
+int sys_getcwd(char *buf, size_t size);
+
+char* getcwd(char *buf, size_t size) {
+    if (buf) {
+        if (!size) {
+            errno = EINVAL;
+            return (char *) NULL;
+        }
+    }
+
+    if (!buf) {
+        if (size == 0) {
+            size = 1024;
+        }
+        buf = (char *) malloc(size);
+    }
```

Comment:
Would it be fine to just return `EINVAL` on `!size` and `ENOMEM` on `!buf`?

Or do applications actually depend on this unspecified behavior of getcwd allocating for them?

Reply:
I don't know, it is specific to GNU GLIBC according to the man page: https://man7.org/linux/man-pages/man3/getcwd.3.html

Since I followed the LibC manpage and not posix spec, I implemented this additional malloc.

Reply:
Apparently, POSIX allows it:

> On some implementations, if buf is a null pointer, getcwd() may obtain size bytes of memory using [malloc()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/malloc.html). In this case, the pointer returned by getcwd() may be used as the argument in a subsequent call to [free()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/free.html). Invoking getcwd() with buf as a null pointer is not recommended in conforming applications.

It seems musl also allocates a buffer if the provided one is null.
https://github.com/kraj/musl/blob/kraj/master/src/unistd/getcwd.c

Reply:
I initially leaned towards not allocating memory at all, but if you think it makes sense to add this preemptively instead of waiting to see if this is even triggered by applications, this is fine by me. :)

It would make sense to use `PATH_MAX`, though, instead of the constant, right?